### DISCUSSION TITLE
Why so many issues replicating data to other DBs with CDC tools?

### DISCUSSION THREAD

**Discussion by BoscoDomingo** [supabase/supabase#20925] [Questions] [answered]
I need to replicate all tables in a schema periodically to BigQuery using CDC tools via Logical Replication, but none have worked thus far, despite having users with the `REPLICATION` permission. The [only article](https://supabase.com/docs/guides/database/replication) I've found on this is not useful whatsoever, and I also [posted this question in the subreddit](https://www.reddit.com/r/Supabase/comments/1afl7sb/does_supabase_support_replication/), Discord, and have asked Supabase Support but to no avail.

This is a must-have for us, otherwise we'll have to switch DB providers and I'd hate that.
<details>
  <summary>Update log</summary>
  
**Update 1**: ~I have managed to get [Stacksync](https://supabase.com/partners/integrations/stacksync) working. We'll evaluate if it's a viable option for us~. However, I still do have grave concerns for the lack of support for Postgres connectors from the vast majority of Data Engineering/Pipeline companies. If Supabase is truly a Postgres wrapper/alternative, it should work out of the box with these IMHO.

---

**Update 2**: Stacksync failed when trying to set up a sync with a more complex table. I've talked to their CEO and they're working on it. In the meantime, there's currently no viable option for us. This is exasperating and we're heavily weighing migrating away from Supabase altogether. How can there be no official way to simply replicate data somewhere else?!?

---

**Update 3**: Thus far Fivetran, Stitch, Google Datastream and Data Fusion, Estuary have all outright failed at one point or another, and Stacksync is still in the air. Support told me

> "The only current limitation with logical replication is that you cannot directly create a subscription within supabase yourself or a user with replication rights (at this moment). Both permissions are in the roadmap"

but I'm not even sure how that relates to this issue. I've asked for further clarification

---

**Update 4**: Airbyte worked with the xmin option for detecting changes, not Logical Replication, which is not enough for us (deletes aren't detected). The majority of Postgres connectors (which should work) don't, and it's extremely likely to do with Logical Replication, or so I think. For now I'll pray and wish that the Supabase team takes a look at why the majority of tools fail to at least rule out the possibility of it being Supabase's or my own fault. For now this issue remains unresolved.

---

**Update 5**: Still no answer to this. Google Data Fusion failed with the JDBC driver (it didn't like the `DECIMAL`/`NUMERIC` type) and Stitch failed while extracting. Again, all issues arise when connecting to Supabase or performing operations on it, no matter the plugin used for the replication slot (`pg_output` or `wal2json`). At this stage we're likely giving up on this, implementing a hacky solution to buy some time in hopes you'll implement a fix or provide an answer at some point. Otherwise we'll be forcibly migrating away from Supabase when the workaround no longer suffices. 

---

**Update 6**: @encima has been working closely with us to get this solved. Thus far still nothing working for us, but **he's proven this is indeed possible** with his own PoCs, which is a huge step forward. 
It makes sense that our DBs are therefore misconfigured or, at least, misbehaving. We'll dive into it and report back.

On a side note, here's some useful info from our convo (for those searching on Google and whatnot):

- Supabase does support replication FROM Supabase TO an external destination
- Supabase does not (currently) support replication FROM an external source TO Supabase (using subscriptions)
- By default, the `postgres` role has the replication privilege and no other users
- Supabase supports replication slots of type `wal2json` and `pgoutput`
- You must use the direct connection string and not the connection pooler
- Using the tools you mentioned [various 3rd party tools], you do not need subscriptions as the external tool is the subscriber and Supabase is the publisher

---

**Update 7**: Solved. The issues mostly stemmed from the recent IPv6 migration and 3rd parties not supporting it in their connectors yet. We were also using the connection pooler host in session mode, which is a mistake. You must use the direct connection URL (the "old" one). Uncheck this box to get it:
![image](https://github.com/supabase/supabase/assets/46006784/8ee4a9a2-60ad-4769-9b7d-e286f4a9213d)

More details in [this comment](https://github.com/orgs/supabase/discussions/20925#discussioncomment-8453102)
  
</details>

TL;DR: Supabase Replication via 3rd party CDC tools IS possible, but requires an IPv4 ($4/mo) + at least Pro plan  ($25/mo) until IPv6 is supported by 3rd parties. There's also a few things to keep in mind per tool as stated in the answer. Check the update log above for details on this entire process.

**Comment by BoscoDomingo** [comment]
@kiwicopple @burggraf Pinging you here because I believe this matter is important enough (the vast majority of data pipelines rely on this being supported to work) and I have, after 4 days on this topic, not been able to find a single confirmation on the status of this matter. I've looked on Reddit, Discord, here and over email. Nothing.

**This is a must at my current employer and we'd leave Supabase should it not be possible.**
Thus far Fivetran, GCP Datastream and Estuary (all via their PostgreSQL connectors) haven't worked despite following their guides to a T.
I'm going to keep trying, but it's getting ridiculous - especially having to ask support for basic roles. I get not giving `SUPERUSER` access, but it's overly tedious to have to go back and forth. Please consider automating this process -.

Feel free to ping other people from the team who may be better-equipped to answer this question. Thanks in advance

**Reply by encima** [reply]
Creating a user with replication privileges is coming soon. The postgres user you get has replication privileges though and it is possible to use that user with tools like Fivetran and Airbyte.

**Reply by BoscoDomingo** [reply]
Yes, it *should* be possible. I totally agree. So why isn't it? I can provide project IDs and schemas privately, in case it could have something to do with data types or permissions.

**Reply by encima** [reply]
It is possible  ([docs](https://github.com/supabase/supabase/blob/d7dd4eeca6d689e32cb7fd1d04f1febe6f2f8a7e/apps/docs/pages/guides/database/postgres/setup-replication-external.mdx) awaiting merge), we have a number of customers using Fivetran and Airbyte. As far as I know, AWS DMS does not support this because it needs some additional event triggers beyond the scope of logical replication but others do work (my own projects are using Airbyte currently without additional roles needed)
Not being able to use a separate role is not an optimal experience and we aim to resolve that soon.

If you are not able to get it working, please open a Support ticket and ask for it to be assigned to Chris on the Success team and link to this discussion.

Note: you do not need to ask for additional roles or commands to be run by Support. All* of the necessary actions are possible with a fresh project with the `postgres` user

* aside from modifying settings like `wal_keep_size`

**Reply by BoscoDomingo** [reply]
> It is possible ([docs](https://github.com/supabase/supabase/blob/d7dd4eeca6d689e32cb7fd1d04f1febe6f2f8a7e/apps/docs/pages/guides/database/postgres/setup-replication-external.mdx) awaiting merge)

What you're doing there is a replication to another Postgres instance. I am sending it to BigQuery, not sure it applies since we'd have to use service accounts instead of DB users.
Subscriptions aren't really mentioned anywhere in your docs either (first time seeing the `CREATE SUBSCRIPTION` command).

> we have a number of customers using Fivetran and Airbyte

Using Logical Replication? That would be nice to know, and if so, how. Because they both failed during setup with their Postgres connectors for us despite following their guides.

> If you are not able to get it working, please open a Support ticket and ask for it to be assigned to Chris on the Success team and link to this discussion.

I already have a thread open with support, so I'll add what you said there, thanks.
Fivetran isn't really what we want to go with (supposedly it's extremely expensive and we want to avoid having to switch again in the near term due to that) so we'll try to make it work with any of the other CDC tools that have failed.

> Note: you do not need to ask for additional roles or commands to be run by Support. All* of the necessary actions are possible with a fresh project with the postgres user

We already have a user with replication permissions, as we want to avoid using `postgres` for everything, so not an issue.

**Reply by encima** [reply]
I am, although this is the same for any tool. For any tool that works with Postgres, we support those that support logical replication. 
Replicating from PG to any X is using a publication only, BigQuery is the subscriber in your case so this should be working with them (if BQ cannot handle using an existing replication slot and tries to make them each time then this is an issue with BQ instead of Supabase)

From this:
- I will try and setup BQ replication to see the issues (using Datastream)
- What failures did you see with Fivetran (I understand you would not like to use them due to cost reasons but purely to try and replicate the issues on our Staging env)

**Reply by BoscoDomingo** [reply]
Fivetran error:
![image](https://github.com/supabase/supabase/assets/46006784/986f2a61-1b69-441b-acf2-edc32f0ab2a6)

Even though their docs say that if this query runs, it should all be ok
![image](https://github.com/supabase/supabase/assets/46006784/0d001b15-7653-406c-b916-2f07e6321798)

**Reply by BoscoDomingo** [reply]
Google Datastream error:
```
syntax error at or near "START_REPLICATION"
LINE 1: START_REPLICATION SLOT "test_slot_v1" LOGICAL 0/00000000 ("p...
        ^
```
I've tried creating a different replication slot, no luck. Attempted with both users that have REPLICATION permissions as well

**Comment by mwoss** [comment]
High chance that you've probably tried it, but have you checked official Google docs on [Postgres replication to BigQuery](https://cloud.google.com/data-fusion/docs/how-to/reading-from-postgresql)? Underneath, Supabase is just a Postgres, so you should be able to do all sort of things on it, like with a normal Postgres cluster.

**Reply by BoscoDomingo** [reply]
I hadn't seen Google Data Fusion, no, so I'll give it a go today and report back.

And I agree that Supabase should just work with any Postgres connector. That's the main reason why it's so incredibly frustrating that it doesn't, and the source of my concern.

**Reply by BoscoDomingo** [reply]
@mwoss Failure as well. This time with a more specific reason than other tools, but something I can't fix: 
![image](https://github.com/supabase/supabase/assets/46006784/d84f843e-57d7-45bb-b714-101f5829fa47)

I do use DECIMAL in my tables as I deal with currencies, so that cannot be changed, but again, another postgres driver that simply doesn't work with Supabase. It's just unbelievable at this point

**Reply by mwoss** [reply]
It seems decimal types in BigQuery need to be handled differently from how they are typically represented in Postgres. I would not say it's a Postgres/Supabase issue tho. I looked up and in BigQuery, decimal types are represented using the `NUMERIC` data type, so you might need to handle the conversion explicitly as Google CDF is not able to apply conversion automatically.

**Reply by BoscoDomingo** [reply]
That still doesn't explain the issues connecting from all the other tools. Fivetran for example fails when validating the database setup, before I have selected BigQuery as the destination. This is a Data-Fusion-specific issue. Also, how would I do that if the connector itself is failing before I can do any transformations? See my frustration with this whole process, which should've been a walk in the park?

**Reply by BoscoDomingo** [reply]
[The issue is with the driver](https://github.com/cdapio/cdap/blob/c2743d8d69cd8538f579fba9a7f684ede32ca33e/cdap-api-common/src/main/java/io/cdap/cdap/api/data/schema/Schema.java#L645). As you can imagine, not gonna wait for GCP to implement their solution to this...

**Reply by mwoss** [reply]
Undestable, this has to be frustrating. Btw have you tried maybe managed solutions like Airbyte? Maybe they will work in your case.

**Reply by BoscoDomingo** [reply]
It's the next one in line, coincidentally. Let's see if that one works :/

**Reply by mwoss** [reply]
Finger crossed, hope it gonna work 🤞

**Reply by BoscoDomingo** [reply]
@mwoss Airbyte failed too, shocker.
![image](https://github.com/supabase/supabase/assets/46006784/e57a5566-8229-4f04-b9d2-2e83b7308b96)

**Reply by BoscoDomingo** [reply]
Managed to connect via xmin instead of Logical replication, but it's now failing on BigQuery's end, so I can't test if the pipeline would work. That being said, Airbyte may be the solution if someone were to use a different destination and didn't mind not using logical replication
![image](https://github.com/supabase/supabase/assets/46006784/c68f19b1-7353-4829-b247-2d618e630307)

**Reply by mwoss** [reply]
Oh :< Btw just to double check if it's not a Supabase issue, have you tried to set up a separate Postgres instance with the same schema but a sample of your data and then replicate it to BigQuery? I still think it's a tooling issue not a Supabase, but could be wrong tho.

**Reply by BoscoDomingo** [reply]
Turns out the BigQuery error it was me being a dum dum after so long just banging my head against multiple walls. Seems like it might be working now.

Edit: Indeed, Airbyte worked. A bit cumbersome how it adds a few columns but nothing I can't easily hide in queries. Damn that was annoying. I'll update the post

**Reply by mwoss** [reply]
Great :3 Good to hear that it works now

**Reply by BoscoDomingo** [reply]
Not really. It unfortunately uses xmin, which is to say, I can implement that myself just dumping data that was updated past a certain timestamp. It's not particularly sophisticated (deletes aren't synced) and is different from Logical Replication. The issue with Supabase persists

**Comment by BoscoDomingo** [comment]
For future reference: Airbyte with xmin as the CDC method worked, but isn't what we're looking for. Logical replication simply didn't. I'll wait to hear back from Stacksync to see if theirs is a viable option.

None of the other tools I tried worked either. This should be looked at by someone at Supabase at some point if they intend for this service to work with data pipelines.
I am lucky enough to be able to pick any service, but someone who is already locked to one of the ones that didn't work may not find this amusing. I sure don't...

**Comment by BoscoDomingo** [comment]
FYI This is the Stitch error (uses `wal2json` instead of `pgoutput` as the plugin for the replication slot):
![image](https://github.com/supabase/supabase/assets/46006784/500a51de-ae3d-4721-98a1-c7b5df676c95)

**Reply by BoscoDomingo** [reply]
Double-checked: The initial load works, but changes aren't streamed from that point onwards. Clearly something's stopping them from doing so

**Reply by encima** [reply]
`wal2json` is used in the Supabase `realtime` stack and you can create a replication slot that uses it:
`SELECT pg_create_logical_replication_slot('example_wal_slot', 'wal2json');`

This is an issue with Stitch (or a feature request) to support other formats for slots. I am unsure but I assume Stitch allows subscribing to an existing replication slot?

**Reply by encima** [reply]
@BoscoDomingo  just tested with Stitch using the command above to create a replication slot with `wal2json` and this was successful.
The only command I needed to run was for the replication slot (and creating a destination database). I will leave it running with some bg usage and report back if the same issues occur

**Reply by encima** [reply]
Update: Stitch is confirmed working (if 2 hours can be called confirmed 😅 ). The initial load was successful and subsequent extractions have worked.
At the very least, this indicates that we need docs on setting this up.

**Reply by BoscoDomingo** [reply]
@encima Interesting, I couldn't get extractions working at all (I created the logical replication slot with `wal2json`) and initial load worked but no changes from that point onwards. I tried with 2 different tables in 2 different schemas.

**Reply by BoscoDomingo** [reply]
Yup, confirmed extractions are still not working for me for neither table. User has replication permissions and one of them is in the public schema with 4 rows. Changes are detected in Supabase, but Stitch cannot read them
![image](https://github.com/supabase/supabase/assets/46006784/47b5f333-3613-46cf-8034-f8404b73886d)
![image](https://github.com/supabase/supabase/assets/46006784/12b5975f-254d-4209-bcf2-4cb6f1dd9fa0)

**Reply by BoscoDomingo** [reply]
![image](https://github.com/supabase/supabase/assets/46006784/e1d3a1e8-d280-4e75-8128-ca9c0eaf8bf2)

It only detects 2495 rows when it's up to ~2530

**Reply by BoscoDomingo** [reply]
@encima I assume you're not doing this with SUPERUSER access, right? It's the one requirement from the docs I can't fulfil
![image](https://github.com/supabase/supabase/assets/46006784/37253a13-5fe9-4673-92f1-c8ec2ac555c4)

**Reply by encima** [reply]
I am not. I am using the standard postgres user from a fresh project

**Reply by encima** [reply]
Superuser is only needed to modify settings (and some could be optimised). It is odd though; does Stitch report errors on any of the extractions it has run?

**Reply by BoscoDomingo** [reply]
Yes, on every single extraction, including the first one where it does take all the rows that existed at that point as I mentioned. From then onwards, changes aren't streamed (replication cannot be started)

**Comment by BoscoDomingo** [comment]
This is the error in Estuary (tried with 2 users, none worked):
![image](https://github.com/supabase/supabase/assets/46006784/3f20a04f-67de-498d-aa3e-3fd5145d1d5a)

**Reply by encima** [reply]
<img width="1294" alt="image" src="https://github.com/supabase/supabase/assets/517923/c63456a5-133f-4967-aa8e-2a377e61f1f1">

Configuration working with Estuary
- Precreate publication for all tables (since you will filter it in Estuary anyway)
- Precreate replication slot (using `pgoutput`)
- Fun tip: maybe don't use sqlite as the destination (as I did) since the max size is 500mb 😅. Would be cool if Estuary supported /dev/null

<img width="478" alt="image" src="https://github.com/supabase/supabase/assets/517923/edf543c1-8029-4089-8b01-3a91933506c9">

**Reply by BoscoDomingo** [reply]
That's crazy, but I'm so happy to know that it is possible. Thank you for the help thus far, we'll keep looking into this.

**Reply by jgraettinger** [reply]
👋 Estuary engineer here. This _used_ to work just fine, but I can definitely see now that the garden path would tend to have users plugging in the connection pooler string -- which isn't going to work.

I've also flagged a potential issue with IPv6-only databases and our connector that we'll track down.

@encima Did you find it was _required_ for some reason to pre-create the publication group and replication slot when using the `postgres` user ? When using a user with sufficient privilege, our connector will do that automatically just to make "hello world" use cases more seamless.

**Reply by encima** [reply]
> 👋 Estuary engineer here. This _used_ to work just fine, but I can definitely see now that the garden path would tend to have users plugging in the connection pooler string -- which isn't going to work.
> 
> 
> 
> I've also flagged a potential issue with IPv6-only databases and our connector that we'll track down.
> 
> 
> 
> @encima Did you find it was _required_ for some reason to pre-create the publication group and replication slot when using the `postgres` user ? When using a user with sufficient privilege, our connector will do that automatically just to make "hello world" use cases more seamless.

@jgraettinger No reason to pre-create it; my script just does that anyway for all platforms I tested. 

Would be happy to work on a single doc that we can publish on both platforms if it's easier

**Comment by encima** [comment]
Adding a summary from the open support ticket:
- We will keep discussions here so that the resolution is public (sorry for the notification spam for some, feel free to unsubscribe)
- Please do not ping those not involved in the discussion for visibility; GH notifications are tough enough as it is! (feel free to ping me if I have not responded for a few days)
- Let's use max one thread per tool (unless it is 2 different setups)
- When posting the issue for each tool, let's make sure that it can be reproduced as much as possible
    - use the `postgres` user
    - if possible, post a minimal reproducible Supabase DB (schema or full dump if small and not sensitive)
    - Outline the setup of each step (or record it, if you prefer)
- The aim is to identify 
    - tools that do/do not work 
    - documentation gaps
    - permissions and commands that can be made easier

The current status is: Supabase does support logical replication with the postgres user (replicating to external tools) and we need to update our documentation to reflect those that work/do not work.

Massive thanks to @BoscoDomingo for the heroic debugging effort they have made so far and for their patience, let's solve this for you and everyone 💪

**Reply by BoscoDomingo** [reply]
Thank you. As mentioned via email, we'll coordinate to discard the possibility of an issue with our DBs and pinpoint the exact reason for the weird bugs. I'm also updating the OP to reflect this

**Reply by encima** [reply, marked as answer]
Closing this one out, there are 3 things to note when setting up replication:

1. Does the connector support IPv6? This is not always easy to know but you will usually see connection or hostname issues when testing. We have the [IPv4 addon](https://supabase.com/docs/guides/platform/ipv4-address) if the system you want to use does not yet support it

2. Use the direct connection (not the pooler) in every case; logical replication needs this. We do have this outlined in the [Replication docs](https://supabase.com/docs/guides/database/postgres/setup-replication-external) and we will work to provide specific instructions for different systems in the docs soon

3. When creating a replication slot, make sure it is using the format expected by the system. Some will require `pgoutput` or `wal2json` and it is not always clear so make sure to check this with the system you are using

With that, it looks like this is "solved" in the context of the original ask with the follow up task for us to update documentation with example setups for some systems, such as Estuary or Stitch or Airbyte. If you work at (or with) a system that you want us to include, please contact us via Support.

**Reply by BoscoDomingo** [reply]
For future reference: 
- we were admittedly making the mistake of using the connection pooler string in Session mode (port 5432)
- not having an IPv4 address (which we purchased for $4/month as mentioned) caused issues with connectors, even when they were able to resolve the address, none could connect. Until that is supported by each tool, assume you'll need an IPv4 (**which also implies paying for the Pro plan at minimum**, so $29/mo at least if you want DB replication).

Thank you for the awesome work @encima and hope this serves for future users who may come across this issue. I'll update the OP to reflect this resolution 👍🏼

**Comment by njoshi22** [comment]
I struggled for 10 hours today trying to get Debezium set up. For future reference, you should update your docs to make clear to use a separate role with `replication` permissions - using the default postgres role did not work.

**Reply by encima** [reply]
The postgres role has replication permissions. We will recommend using a separate user number currently do not support giving replication permissions to other users
